File: src/geometry/sphere.cpp
#include "sphere.h"
#include <cmath>

Sphere::Sphere(const Point& center,
               float radius,
               std::shared_ptr<Material> material)
{
  this->center = center;
  this->radius = radius;
  this->material = material;
}

std::optional<Observation>
Sphere::hit(const Ray& ray, float tMin, float tMax) const
{
  Vec oc = ray.origin - center;

  float a = ray.direction.lenSquared();
  float halfB = ray.direction.dot(oc);
  float c = oc.lenSquared() - radius * radius;

  float delta = halfB * halfB - a * c;
  if (delta < 0) {
    return {};
  }

  float sqrtD = sqrt(delta);
  float root = (-halfB - sqrtD) / a;

  if (root < tMin) {
    root = (-halfB + sqrtD) / a;
    if (root < tMin || tMax < root) {
      return {};
    }

  } else if (root > tMax) {
    // This is the most negative root, if it's already too big
    // then the other root will be bigger too
    return {};
  }

  Observation obs;
  obs.t = root;
  obs.point = ray.at(root);
  obs.material = material;

  Vec outwardNormal = (obs.point - center) / radius;
  obs.setFace(ray, outwardNormal);

  return obs;
}

File: src/geometry/sphere.rs
use super::*;
use bounding::BoundingBox;
use material::Material;
use std::sync::Arc;
use transform::Transform;

#[derive(Clone)]
pub struct Ellipsoid {
    transform: Transform,
    material: Arc<dyn Material>,
    bounds: BoundingBox,
}

impl Ellipsoid {
    /// Builds a sphere given a center and a radius
    /// Will error out if the radius is 0
    pub fn sphere(center: Point, radius: Float, material: Arc<dyn Material>) -> Arc<Self> {
        let radius_vec = radius * vector::ONES;
        let transform = Transform::scale_all(1.0 / radius) * Transform::make_origin(center);

        Arc::new(Self {
            transform,
            material,
            bounds: BoundingBox::from_extrema(center - radius_vec, center + radius_vec),
        })
    }

    /// Builds an axis-aligned ellipsoid given a center and a vector of semiaxis lengths.
    /// Will error out if any length is 0
    pub fn new(center: Point, semiaxes: Vector, material: Arc<dyn Material>) -> Arc<Self> {
        let transform = Transform::scale(1.0 / semiaxes.x, 1.0 / semiaxes.y, 1.0 / semiaxes.z)
            * Transform::make_origin(center);

        Arc::new(Self {
            transform,
            material,
            bounds: BoundingBox::from_extrema(center - semiaxes, center + semiaxes),
        })
    }
}

impl Geometry for Ellipsoid {
    fn collide(&self, ray: &Ray, t_range: Range) -> Option<Collision> {
        let (original_ray, ray) = (ray, self.transform * (*ray));
        let center_origin = (-ray.origin).into();

        let a = ray.direction.norm_squared();
        let h = ray.direction.dot(&center_origin);
        let c = center_origin.norm_squared() - 1.0;

        let delta = h * h - a * c;
        if delta < 0.0 {
            return None;
        }

        let delta_sqrt = delta.sqrt();
        let mut root = (h - delta_sqrt) / a;
        if t_range.not_contains(root) {
            root = (h + delta_sqrt) / a;
            if t_range.not_contains(root) {
                return None;
            }
        }

        let glancing_point = ray.at(root);
        let mut collision = Collision {
            point: glancing_point,
            normal: glancing_point.into(),
            t: root,
            is_front_facing: true,
            material: self.material.clone(),
        };
        collision.apply(&original_ray, self.transform);

        Some(collision)
    }

    fn bounding_box(&self) -> BoundingBox {
        self.bounds
    }
}
